### PR TITLE
[Merged by Bors] - feat: Uniformly and strongly convex functions

### PR DESCRIPTION
**Pull request by chenyili0818**
We define uniform convex functions with modulus `φ` and `m`-strongly convex functions on a normed space.

Co-authored-by: Ziyu Wang <tropicalfish910@gmail.com>
Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

---

This is a modified version after receiving the response in https://github.com/leanprover-community/mathlib4/pull/7011

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ADedecker** [comment]
Could you import your new file in `Mathlib.lean`. Once this is done and the PR is ready for review, please tag it with `awaiting-review`, otherwise reviewers and maintainer might miss it!

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  f (x) - m * ‖x‖ ^ 2 / 2 is a convex function -/
+def StronglyConvexOn (s : Set E) (m : ℝ) (f : E → ℝ) : Prop :=
+ m > 0 ∧ ConvexOn ℝ s (fun x ↦ f x - m / 2 * ‖x‖ ^ 2)
+
+variable (s : Set E) {f : E → ℝ}
+
+lemma equiv_lemma {x y : E} {o m: ℝ} (h₁ : o > 0) (h₂ : o < 1):
+o * (f x - m / 2 * ‖x‖ ^ 2) + (1 - o) * (f y - m / 2 * ‖y‖ ^ 2)
+  + m / 2 * ‖o • x + (1 - o) • y‖ ^ 2 =
+o * f x + (1 - o) * f y - m / 2 * o * (1 - o) * ‖x - y‖ ^ 2 :=
```
Comment (ADedecker) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:36]: Could you check that your code follows our [style recommandations](https://leanprover-community.github.io/contribute/style.html). In this case, the statement of the lemma should be indented twice. Thanks!

Comment (chenyili0818) [reply Mathlib/Analysis/Convex/StronglyConvex.lean:36]: Thank you for your advice and I will fix it and push on github again

**Comment by jcommelin** [comment]
FYI: The `easy` label is for PRs that can be reviewed in < 20s of review time.

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+θ * f x + (1 - θ) * f y - m / 2 * θ * (1 - θ) * ‖x - y‖ ^ 2 := by
+  rcases strongly_convex with ⟨hm, _, hh⟩
+  constructor; exact hm
+  intro x hx y hy o hg hl
+  specialize @hh x hx y hy o (1 - o) (by linarith) (by linarith) (by ring)
+  simp only [smul_eq_mul, tsub_le_iff_right] at hh
+  rw [equiv_lemma hg (by linarith)] at hh
+  exact hh
+
+theorem Strongly_Convex_Def (hs : Convex ℝ s) (m : ℝ) (hm : m > 0)
```
Comment (YaelDillies) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:60]:
```suggestion
theorem stronglyConvexOn_def (hs : Convex ℝ s) (hm : m > 0)
```

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    x ∈ s → ∀ (y : E) , y ∈ s → ∀ (θ : ℝ), θ > 0 → θ < 1
+      → f (θ • x + (1 - θ) • y) ≤ θ * f x + (1 - θ) * f y - m / 2 * θ * (1 - θ) * ‖x - y‖ ^ 2 := by
+  constructor
+  · intro h
+    exact Strongly_Convex_Bound s m h
+  · rintro ⟨hm, h'⟩
+    exact Strongly_Convex_Def s hs m hm h'
+
+lemma Strongly_convex_is_strictlyconvex (m : ℝ) (hsc: StronglyConvexOn s m f):
+StrictConvexOn ℝ s f := by
```
Comment (YaelDillies) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:98]:
```suggestion
lemma StronglyConvexOn.strictConvexOn (hsc : StronglyConvexOn s m f) : StrictConvexOn ℝ s f := by
```

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+Add derivative information for a strongly convex functions, and prove some relevant properties.
+Connect with L_smooth property and prove more theorems.
+-/
+
+variable {n : Type _} [Fintype n]  {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E]
+
+/-- Function f (x) is m - strongly convex iff
+  f (x) - m * ‖x‖ ^ 2 / 2 is a convex function -/
+def StronglyConvexOn (s : Set E) (m : ℝ) (f : E → ℝ) : Prop :=
+ m > 0 ∧ ConvexOn ℝ s (fun x ↦ f x - m / 2 * ‖x‖ ^ 2)
```
Comment (YaelDillies) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:29]: I would not enforce `0 < m` as part of the definition
```suggestion
  ConvexOn ℝ s (fun x ↦ f x - m / 2 * ‖x‖ ^ 2)
```
(at least try this and see whether it's nicer)

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+variable (s : Set E) {f : E → ℝ}
+
+lemma equiv_lemma {x y : E} {o m: ℝ} (h₁ : o > 0) (h₂ : o < 1):
+o * (f x - m / 2 * ‖x‖ ^ 2) + (1 - o) * (f y - m / 2 * ‖y‖ ^ 2)
+  + m / 2 * ‖o • x + (1 - o) • y‖ ^ 2 =
+    o * f x + (1 - o) * f y - m / 2 * o * (1 - o) * ‖x - y‖ ^ 2 :=
```
Comment (YaelDillies) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:36]: We rather use `0 < a`, `0 < b`, `a + b = 1` than `0 < θ`, `θ < 1`
```suggestion
lemma equiv_lemma {x y : E} {a b m: ℝ} (ha : 0 < a) (hb : 0 < b) (hab : a + b = 1) :
    a * (f x - m / 2 * ‖x‖ ^ 2) + b * (f y - m / 2 * ‖y‖ ^ 2) + m / 2 * ‖a • x + b • y‖ ^ 2
      = a * f x + b * f y - m / 2 * a * b * ‖x - y‖ ^ 2 :=
```

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  exact hfun x hx y hy o hog hol
+
+/-
+  m - strongly convex function ↔ ∀ x, y in domain s, ∀ θ ∈ (0, 1),
+    f (θ • x + (1 - θ) • y) ≤ θ * f x + (1 - θ) * f y - m / 2 * θ * (1 - θ) * ‖x - y‖ ^ 2
+-/
+theorem Strongly_Convex_Iff (hs : Convex ℝ s) (m : ℝ):
+  StronglyConvexOn s m f ↔ m > 0 ∧ ∀ (x :E),
+    x ∈ s → ∀ (y : E) , y ∈ s → ∀ (θ : ℝ), θ > 0 → θ < 1
+      → f (θ • x + (1 - θ) • y) ≤ θ * f x + (1 - θ) * f y - m / 2 * θ * (1 - θ) * ‖x - y‖ ^ 2 := by
```
Comment (YaelDillies) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:90]: Same here
```suggestion
theorem stronglyConvexOn_def :
  StronglyConvexOn s m f ↔ Convex ℝ s ∧ ∀ ⦃x⦄, x ∈ s → ∀ ⦃y⦄, y ∈ s →
     ∀ ⦃a⦄, 0 < a → ∀ ⦃b⦄, 0 < b → a + b = 1 → f (a • x + b • y)
       ≤ a * f x + b * f y - m / 2 * a * b * ‖x - y‖ ^ 2 := by
```

File: Mathlib/Analysis/Convex/StronglyConvex.lean [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  rcases Strongly_Convex_Bound s m hsc with ⟨hm, hfun⟩
+  specialize hfun xm hxm xm' hxm' (1 / 2) (by norm_num) (by norm_num)
+  rw [this, ← add_mul, add_sub, add_comm (1 / 2), ← add_sub, sub_self, add_zero, one_mul] at hfun
+  have hl : f (((1 : ℝ) / 2) • xm + ((1 : ℝ) - 1 / 2) • xm') < f xm' := by
+    have h₂ : m / 2 * (1 / 2) * (1 - 1 / 2) * ‖xm - xm'‖ ^ 2 > 0 := by
+      apply mul_pos
+      · show 0 < m / 2 * (1 / 2) * (1 - 1 / 2)
+        apply mul_pos _ (by linarith)
+        · apply mul_pos _ (by linarith)
+          exact div_pos (by linarith) (by norm_num)
```
Comment (YaelDillies) [inline Mathlib/Analysis/Convex/StronglyConvex.lean:139]:
```suggestion
      · positivity
```

**Comment by chenyili0818** [comment]
Thank you for all your advice. I fix my lean project and renew it here.

**Comment by sgouezel** [comment]
Once you have dealt with a comment, could you please resolve the corresponding conversation to help us see what remains to be done?

**Comment by YaelDillies** [comment]
To let you know, I'm reworking all this so that proofs are quicker and hold in more generality.

**Comment by YaelDillies** [comment]
It now lives at https://github.com/YaelDillies/LeanCamCombi/blob/main/LeanCamCombi/Mathlib/Analysis/Convex/Strong.lean. I need to open a few prerequisite PRs.

**Review by YaelDillies — APPROVED** [review]
I'm happy with the current state and nobody seems to have complained, so 🤷 

maintainer merge

**Comment by chenyili0818** [comment]
Thank you for your help for coding! How can I merge this pull request into the main branch?

**Comment by YaelDillies** [comment]
A maintainer needs to approve it! I've sent [a message on Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.237040.20Uniformly.20and.20strongly.20convex.20functions).

File: Mathlib/Analysis/Convex/Strong.lean [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    using hf.2 hx hy ha hb hab
+
+lemma UniformConvexOn.sub (hf : UniformConvexOn s φ f) (hg : UniformConcaveOn s ψ g) :
+    UniformConvexOn s (φ + ψ) (f - g) := by simpa using hf.add hg.neg
+
+lemma UniformConcaveOn.sub (hf : UniformConcaveOn s φ f) (hg : UniformConvexOn s ψ g) :
+    UniformConcaveOn s (φ + ψ) (f - g) := by simpa using hf.add hg.neg
+
+/-- A function `f` from a real normed space is `m`-strongly convex if
+`f (t • x + (1 - t) • y) ≤ t • f x + (1 - t) • f y - t * (1 - t) * φ ‖x - y‖` for all `t ∈ [0, 1]`.
```
Comment (apurvanakade) [inline Mathlib/Analysis/Convex/Strong.lean:116]: I think you mean "`x ↦ f x - m / 2 * ‖x‖ ^ 2` is convex" is the definition and it is equivalent to the uniformly convex in an inner product space.

File: Mathlib/Analysis/Convex/Strong.lean [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+In an inner product space, this is equivalent to `x ↦ f x - m / 2 * ‖x‖ ^ 2` being convex. -/
+def StrongConvexOn (s : Set E) (m : ℝ) : (E → ℝ) → Prop :=
+  UniformConvexOn s fun r ↦ m / (2 : ℝ) * r ^ 2
+
+/-- A function `f` from a real normed space is `m`-strongly concave if
+`t • f x + (1 - t) • f y + t * (1 - t) * φ ‖x - y‖ ≤ f (t • x + (1 - t) • y)` for all `t ∈ [0, 1]`.
```
Comment (apurvanakade) [inline Mathlib/Analysis/Convex/Strong.lean:123]: Same as above.

File: Mathlib/Analysis/Convex/Strong.lean [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+# Uniformly and strongly convex functions
+
+In this file, we define uniformly convex functions and strongly convex functions.
+
+For a real normed space `E`, a uniformly convex function with modulus `φ : ℝ → ℝ` is a function
+`f : E → ℝ` such that `f (t • x + (1 - t) • y) ≤ t • f x + (1 - t) • f y - t * (1 - t) * φ ‖x - y‖`
+for all `t ∈ [0, 1]`.
+
+A `m`-strongly convex function is a uniformly convex function with modulus `fun r ↦ m / 2 * r ^ 2`.
+If `E` is an inner product space, this is equivalent to `x ↦ f x - m / 2 * ‖x‖` being convex.
```
Comment (apurvanakade) [inline Mathlib/Analysis/Convex/Strong.lean:18]: Typo: `x ↦ f x - m / 2 * ‖x‖ ^ 2`

File: Mathlib/Analysis/Convex/Strong.lean (hunk 8 above)
Comment (YaelDillies) [reply Mathlib/Analysis/Convex/Strong.lean:116]: Your comment doesn't make sense, but so does my docstring! Fixed.

**Comment by ADedecker** [comment]
Thanks!

bors merge

**Pull request closed without merging**